algorithms/maths/rsa.py

```diff
@@ -0,0 +1,93 @@
+"""
+RSA encryption algorithm
+a method for encrypting a number that uses seperate encryption and decryption keys
+this file only implements the key generation algorithm
+
+there are three important numbers in RSA called n, e, and d
+e is called the encryption exponent
+d is called the decryption exponent
+n is called the modulus
+
+these three numbers satisfy
+((x ** e) ** d) % n == x % n
+
+to use this system for encryption, n and e are made publicly available, and d is kept secret
+a number x can be encrypted by computing (x ** e) % n
+the original number can then be recovered by computing (E ** d) % n, where E is
+the encrypted number
+
+fortunately, python provides a three argument version of pow() that can compute powers modulo
+a number very quickly:
+(a ** b) % c == pow(a,b,c)
+"""
+
+# sample usage:
+# n,e,d = generate_key(16)
+# data = 20
+# encrypted = pow(data,e,n)
+# decrypted = pow(encrypted,d,n)
+# assert decrypted == data
+
+import random
+
+
+def generate_key(k, seed=None):
+    """
+    the RSA key generating algorithm
+    k is the number of bits in n
+    """
+
+    def modinv(a, m):
```

Comment:
Use an efficient algorithm like the Extended Euclidean Algorithm to compute modular inverses.

Reply:
Example of possible code:

`def extended_gcd(a, b):
    if b == 0:
        return a, 1, 0
    gcd, x1, y1 = extended_gcd(b, a % b)
    x = y1
    y = x1 - (a // b) * y1
    return gcd, x, y

def mod_inverse(a, m):
    gcd, x, y = extended_gcd(a, m)
    if gcd != 1:
        raise ValueError(f"The modular inverse of {a} modulo {m} does not exist.")
    else:
        return x % m`
        
This code is not perfect due to using recursion, however it allows to boost efficiency.